File: backend/apps/python_agent/rsc_integration.py
"""
Integration module for React Server Components (RSC) with Django.
This module provides functionality for generating and managing React Server Components
based on agent actions and tool usage.
"""

import json
import logging
import uuid
from typing import Any, Dict, List, Optional, Union

from django.conf import settings

from .go_integration import get_go_runtime_integration

logger = logging.getLogger(__name__)

class RSCIntegration:
    """
    Integration class for React Server Components (RSC) with Django.
    """
    
    def __init__(self):
        """
        Initialize the RSC integration.
        """
        self.go_runtime = get_go_runtime_integration()
        self.component_cache = {}
    
    def generate_component(self, component_type: str, props: Dict[str, Any]) -> str:
        """
        Generate a React Server Component.
        
        Args:
            component_type: Type of the component to generate
            props: Properties for the component
            
        Returns:
            Component ID
        """
        try:
            result = self.go_runtime.execute_task(
                task_type="generate_component",
                input_data={
                    "component_type": component_type,
                    "props": props
                },
                description=f"Generate {component_type} component"
            )
            
            if "error" in result:
                logger.error(f"Error generating component: {result['error']}")
                return None
            
            component_id = result.get("component_id")
            if component_id:
                self.component_cache[component_id] = result.get("component", {})
            
            return component_id
        except Exception as e:
            logger.error(f"Error generating component: {e}")
            return None
    
    def generate_component_from_agent_action(
        self, 
        agent_id: str, 
        action_id: str, 
        action_type: str, 
        action_data: Dict[str, Any]
    ) -> str:
        """
        Generate a React Server Component from an agent action.
        
        Args:
            agent_id: ID of the agent
            action_id: ID of the action
            action_type: Type of the action
            action_data: Data for the action
            
        Returns:
            Component ID
        """
        try:
            result = self.go_runtime.execute_task(
                task_type="generate_component_from_agent_action",
                input_data={
                    "agent_id": agent_id,
                    "action_id": action_id,
                    "action_type": action_type,
                    "action_data": action_data
                },
                description=f"Generate component for {action_type} action"
            )
            
            if "error" in result:
                logger.error(f"Error generating component from agent action: {result['error']}")
                return None
            
            component_id = result.get("component_id")
            if component_id:
                self.component_cache[component_id] = result.get("component", {})
            
            return component_id
        except Exception as e:
            logger.error(f"Error generating component from agent action: {e}")
            return None
    
    def generate_component_from_tool_usage(
        self, 
        agent_id: str, 
        tool_id: str, 
        tool_name: str, 
        tool_input: Dict[str, Any], 
        tool_output: Dict[str, Any]
    ) -> str:
        """
        Generate a React Server Component from a tool usage.
        
        Args:
            agent_id: ID of the agent
            tool_id: ID of the tool
            tool_name: Name of the tool
            tool_input: Input data for the tool
            tool_output: Output data from the tool
            
        Returns:
            Component ID
        """
        try:
            result = self.go_runtime.execute_task(
                task_type="generate_component_from_tool_usage",
                input_data={
                    "agent_id": agent_id,
                    "tool_id": tool_id,
                    "tool_name": tool_name,
                    "tool_input": tool_input,
                    "tool_output": tool_output
                },
                description=f"Generate component for {tool_name} tool usage"
            )
            
            if "error" in result:
                logger.error(f"Error generating component from tool usage: {result['error']}")
                return None
            
            component_id = result.get("component_id")
            if component_id:
                self.component_cache[component_id] = result.get("component", {})
            
            return component_id
        except Exception as e:
            logger.error(f"Error generating component from tool usage: {e}")
            return None
    
    def get_component(self, component_id: str) -> Dict[str, Any]:
        """
        Get a component by ID.
        
        Args:
            component_id: ID of the component
            
        Returns:
            Component data
        """
        if component_id in self.component_cache:
            return self.component_cache[component_id]
        
        try:
            result = self.go_runtime.execute_task(
                task_type="get_component",
                input_data={
                    "component_id": component_id
                },
                description=f"Get component {component_id}"
            )
            
            if "error" in result:
                logger.error(f"Error getting component: {result['error']}")
                return None
            
            component = result.get("component", {})
            if component:
                self.component_cache[component_id] = component
            
            return component
        except Exception as e:
            logger.error(f"Error getting component: {e}")
            return None
    
    def list_components(self) -> List[Dict[str, Any]]:
        """
        List all components.
        
        Returns:
            List of components
        """
        try:
            result = self.go_runtime.execute_task(
                task_type="list_components",
                input_data={},
                description="List all components"
            )
            
            if "error" in result:
                logger.error(f"Error listing components: {result['error']}")
                return []
            
            components = result.get("components", [])
            
            for component in components:
                if "id" in component:
                    self.component_cache[component["id"]] = component
            
            return components
        except Exception as e:
            logger.error(f"Error listing components: {e}")
            return []
    
    def get_components_by_agent(self, agent_id: str) -> List[Dict[str, Any]]:
        """
        Get all components for an agent.
        
        Args:
            agent_id: ID of the agent
            
        Returns:
            List of components
        """
        try:
            result = self.go_runtime.execute_task(
                task_type="get_components_by_agent",
                input_data={
                    "agent_id": agent_id
                },
                description=f"Get components for agent {agent_id}"
            )
            
            if "error" in result:
                logger.error(f"Error getting components by agent: {result['error']}")
                return []
            
            components = result.get("components", [])
            
            for component in components:
                if "id" in component:
                    self.component_cache[component["id"]] = component
            
            return components
        except Exception as e:
            logger.error(f"Error getting components by agent: {e}")
            return []
    
    def get_components_by_tool(self, tool_id: str) -> List[Dict[str, Any]]:
        """
        Get all components for a tool.
        
        Args:
            tool_id: ID of the tool
            
        Returns:
            List of components
        """
        try:
            result = self.go_runtime.execute_task(
                task_type="get_components_by_tool",
                input_data={
                    "tool_id": tool_id
                },
                description=f"Get components for tool {tool_id}"
            )
            
            if "error" in result:
                logger.error(f"Error getting components by tool: {result['error']}")
                return []
            
            components = result.get("components", [])
            
            for component in components:
                if "id" in component:
                    self.component_cache[component["id"]] = component
            
            return components
        except Exception as e:
            logger.error(f"Error getting components by tool: {e}")
            return []

_rsc_integration = None

def get_rsc_integration() -> RSCIntegration:
    """
    Get the RSC integration singleton instance.
    
    Returns:
        RSC integration instance
    """
    global _rsc_integration
    if _rsc_integration is None:
        _rsc_integration = RSCIntegration()
    return _rsc_integration

File: pkg/sharedstate/rsc_integration.go
package sharedstate

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"sync"

	"github.com/spectrumwebco/agent_runtime/internal/eventstream"
	"github.com/spectrumwebco/agent_runtime/pkg/ai/vercel"
	"github.com/spectrumwebco/agent_runtime/pkg/sharedstate/adapters"
	"github.com/spectrumwebco/agent_runtime/pkg/sharedstate/models"
)

type RSCIntegration struct {
	sharedStateManager *SharedStateManager
	rscAdapter         *adapters.RSCAdapter
	eventStream        *eventstream.Stream
	rscClient          *vercel.RSCAdapter
	rscIntegration     *vercel.RSCIntegration

	componentCallbacks map[string]func(component vercel.GeneratedComponent)
	mu                 sync.RWMutex
}

func NewRSCIntegration(
	sharedStateManager *SharedStateManager,
	rscAdapter *adapters.RSCAdapter,
	eventStream *eventstream.Stream,
	rscClient *vercel.RSCAdapter,
	rscIntegration *vercel.RSCIntegration,
) *RSCIntegration {
	integration := &RSCIntegration{
		sharedStateManager:  sharedStateManager,
		rscAdapter:          rscAdapter,
		eventStream:         eventStream,
		rscClient:           rscClient,
		rscIntegration:      rscIntegration,
		componentCallbacks:  make(map[string]func(component vercel.GeneratedComponent)),
	}

	integration.setupEventListeners()

	return integration
}

func (i *RSCIntegration) setupEventListeners() {
	if i.eventStream == nil {
		return
	}

	i.eventStream.Subscribe("state_update", func(event eventstream.Event) {
		var stateEvent struct {
			StateType string                 `json:"state_type"`
			StateID   string                 `json:"state_id"`
			Data      map[string]interface{} `json:"data"`
		}

		err := json.Unmarshal([]byte(event.Payload), &stateEvent)
		if err != nil {
			log.Printf("Error unmarshaling state update event: %v", err)
			return
		}

		if stateEvent.StateType == string(models.StateTypeComponent) {
			i.handleComponentStateUpdate(stateEvent.StateID, stateEvent.Data)
		}
	})

	i.eventStream.Subscribe("agent_action", func(event eventstream.Event) {
		var actionEvent struct {
			AgentID    string                 `json:"agent_id"`
			ActionID   string                 `json:"action_id"`
			ActionType string                 `json:"action_type"`
			ActionData map[string]interface{} `json:"action_data"`
		}

		err := json.Unmarshal([]byte(event.Payload), &actionEvent)
		if err != nil {
			log.Printf("Error unmarshaling agent action event: %v", err)
			return
		}

		go func() {
			componentID, err := i.rscAdapter.GenerateComponentFromAgentAction(
				context.Background(),
				actionEvent.AgentID,
				actionEvent.ActionID,
				actionEvent.ActionType,
				actionEvent.ActionData,
			)
			if err != nil {
				log.Printf("Error generating component from agent action: %v", err)
				return
			}

			i.notifyComponentGenerated(componentID)
		}()
	})

	i.eventStream.Subscribe("tool_usage", func(event eventstream.Event) {
		var toolEvent struct {
			AgentID    string                 `json:"agent_id"`
			ToolID     string                 `json:"tool_id"`
			ToolName   string                 `json:"tool_name"`
			ToolInput  map[string]interface{} `json:"tool_input"`
			ToolOutput map[string]interface{} `json:"tool_output"`
		}

		err := json.Unmarshal([]byte(event.Payload), &toolEvent)
		if err != nil {
			log.Printf("Error unmarshaling tool usage event: %v", err)
			return
		}

		go func() {
			componentID, err := i.rscAdapter.GenerateComponentFromToolUsage(
				context.Background(),
				toolEvent.AgentID,
				toolEvent.ToolID,
				toolEvent.ToolName,
				toolEvent.ToolInput,
				toolEvent.ToolOutput,
			)
			if err != nil {
				log.Printf("Error generating component from tool usage: %v", err)
				return
			}

			i.notifyComponentGenerated(componentID)
		}()
	})
}

func (i *RSCIntegration) handleComponentStateUpdate(stateID string, data map[string]interface{}) {
	componentID := stateID
	if len(componentID) > 10 && componentID[:10] == "component:" {
		componentID = componentID[10:]
	}

	component, err := i.rscAdapter.GetComponent(context.Background(), componentID)
	if err != nil {
		log.Printf("Error getting component: %v", err)
		return
	}

	i.notifyComponentUpdated(*component)
}

func (i *RSCIntegration) notifyComponentGenerated(componentID string) {
	component, err := i.rscAdapter.GetComponent(context.Background(), componentID)
	if err != nil {
		log.Printf("Error getting component: %v", err)
		return
	}

	i.notifyComponentUpdated(*component)
}

func (i *RSCIntegration) notifyComponentUpdated(component vercel.GeneratedComponent) {
	i.mu.RLock()
	for _, callback := range i.componentCallbacks {
		go callback(component)
	}
	i.mu.RUnlock()

	if i.eventStream != nil {
		componentJSON, err := json.Marshal(component)
		if err != nil {
			log.Printf("Error marshaling component: %v", err)
			return
		}

		event := eventstream.Event{
			Type:    "component_updated",
			Payload: string(componentJSON),
		}

		err = i.eventStream.Publish(event)
		if err != nil {
			log.Printf("Error publishing component updated event: %v", err)
		}
	}
}

func (i *RSCIntegration) RegisterComponentCallback(id string, callback func(component vercel.GeneratedComponent)) {
	i.mu.Lock()
	defer i.mu.Unlock()

	i.componentCallbacks[id] = callback
}

func (i *RSCIntegration) UnregisterComponentCallback(id string) {
	i.mu.Lock()
	defer i.mu.Unlock()

	delete(i.componentCallbacks, id)
}

func (i *RSCIntegration) GenerateComponent(ctx context.Context, componentType vercel.ComponentType, props map[string]interface{}) (string, error) {
	return i.rscAdapter.GenerateComponent(ctx, componentType, props)
}

func (i *RSCIntegration) GetComponent(ctx context.Context, componentID string) (*vercel.GeneratedComponent, error) {
	return i.rscAdapter.GetComponent(ctx, componentID)
}

func (i *RSCIntegration) ListComponents(ctx context.Context) ([]vercel.GeneratedComponent, error) {
	return i.rscAdapter.ListComponents(ctx)
}

func (i *RSCIntegration) GetComponentsByAgent(ctx context.Context, agentID string) ([]vercel.GeneratedComponent, error) {
	if i.rscIntegration != nil {
		return i.rscIntegration.GetComponentsByAgent(agentID)
	}

	states, err := i.sharedStateManager.ListStates(models.StateTypeAction)
	if err != nil {
		return nil, fmt.Errorf("error listing action states: %w", err)
	}

	componentIDs := make([]string, 0)
	for _, state := range states {
		if state.Data == nil {
			continue
		}

		stateAgentID, ok := state.Data["agent_id"].(string)
		if !ok || stateAgentID != agentID {
			continue
		}

		componentID, ok := state.Data["component_id"].(string)
		if !ok {
			continue
		}

		componentIDs = append(componentIDs, componentID)
	}

	components := make([]vercel.GeneratedComponent, 0, len(componentIDs))
	for _, componentID := range componentIDs {
		component, err := i.rscAdapter.GetComponent(ctx, componentID)
		if err != nil {
			continue
		}

		components = append(components, *component)
	}

	return components, nil
}

func (i *RSCIntegration) GetComponentsByTool(ctx context.Context, toolID string) ([]vercel.GeneratedComponent, error) {
	if i.rscIntegration != nil {
		return i.rscIntegration.GetComponentsByTool(toolID)
	}

	states, err := i.sharedStateManager.ListStates(models.StateTypeTool)
	if err != nil {
		return nil, fmt.Errorf("error listing tool states: %w", err)
	}

	componentIDs := make([]string, 0)
	for _, state := range states {
		if state.Data == nil {
			continue
		}

		stateToolID, ok := state.Data["tool_id"].(string)
		if !ok || stateToolID != toolID {
			continue
		}

		componentID, ok := state.Data["component_id"].(string)
		if !ok {
			continue
		}

		componentIDs = append(componentIDs, componentID)
	}

	components := make([]vercel.GeneratedComponent, 0, len(componentIDs))
	for _, componentID := range componentIDs {
		component, err := i.rscAdapter.GetComponent(ctx, componentID)
		if err != nil {
			continue
		}

		components = append(components, *component)
	}

	return components, nil
}
